terraform/node_resource_plan_destroy.go

```diff
@@ -48,6 +51,36 @@ func (n *NodePlanDestroyableResourceInstance) Execute(ctx EvalContext, op walkOp
 		return diags
 	}
 
+	// Note any upgrades that readResourceInstanceState might've done in the
+	// prevRunState, so that it'll conform to current schema.
+	diags = diags.Append(n.writeResourceInstanceState(ctx, state, prevRunState))
+	if diags.HasErrors() {
+		return diags
+	}
+	// Also the refreshState, because that should still reflect schema upgrades
+	// even if not refreshing.
+	diags = diags.Append(n.writeResourceInstanceState(ctx, state, refreshState))
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// Refresh maybe, except for data resources because it doesn't make sense
+	// to read those when we're just about to discard the result without using
+	// it anyway.
+	if !n.skipRefresh && addr.Resource.Resource.Mode != addrs.DataResourceMode {
```

Comment:
Destroying is done with a separate "refresh plan" beforehand, so I think this is only serving to refresh a second time. The `skipRefresh` behavior during destroy is done by skipping the entire "refresh pan". We need the entire plan graph in order to correctly refresh, because we need to read data sources in case they are required by providers.

Reply:
What seemed to be missing here -- although maybe it was _me_ that was missing something -- is refreshing objects that we are about to destroy, so that e.g. we can detect that they are already gone due to an out-of-band action. From what I observed during testing it seemed like the "normal" plan prior to destroying wasn't refreshing anything that was _only_ tracked in state, and not currently in config.

With that said, it's just as likely that I wasn't testing correctly, so I'll try again and watch more closely.

Reply:
If it's only tracked in state, the plan should be using `NodePlannableResourceInstanceOrphan` which _should_ be able to refresh that state. I think we may run into trouble if the destroy plan graph tries using providers, as it currently just create destroy actions, and mainly exists for targeting to work correctly.

Reply:
Testing again with fresh eyes today what I'm seeing is that it seems to make no difference at all whether this block is here: either way it refreshes only once. Given that, I'm not sure what I observed yesterday but must assume it was a false negative due incorrect testing methodology.

With that said, my changes to `Context.destroyPlan` are saving the `PrevRunState` from the destroy-plan walk rather than from the initial plan walk, because we need that to get updated even when refreshing isn't enabled, and so I think this _does_ need to still save the upgraded state data into the `PrevRunState` because in `-refresh=false` mode this'll be our only opportunity to do that. :thinking:

Reply:
After some playing around, I think the best way to get the necessary behavior in both cases (refresh and not refresh) is to remove this refreshing step you've indicated here, and to make the earlier call to update the refresh state happen only when `n.skipRefresh` is set, in order to compensate for us not having run the earlier plan to cause that to happen.

I don't really like the action at a distance here -- that both the `Context.destroyPlan` function and his instance destroy planning node have to agree on who is responsible for saving the upgraded value into the refresh state and that differs based on `SkipRefresh` -- but it seems like the most reasonable way to get it done without any more risky refactoring.

Reply:
I've pushed up a new version of this which never refreshes in the destroy planning node, and instead it just conditionally writes the upgrade result back to the prev run state and the prior state _only when we're skipping refresh_.

That then coordinates with logic in `Context.destroyPlan` so that:
* In skip refresh mode, the resource instance destroy graph node is responsible for recording the upgrade result
* In normal (refresh enabled) mode, the initial normal plan walk is responsible for recording the upgrade result _and_ the refresh result

I believe that meets my requirement of having all three of the resource instance states conform to current schema when planning succeeds, while avoiding the second refresh my previous attempt was introducing.

Reply:
That action at a distance is how I missed refresh during destroy the last time around. I think the only way to bring these together is to further parameterize the plan graph to create a single complete plan+destroy graph, but that looked like a significant change when I checked last. 

If there was no targeting allowed for destroy, there wouldn't be a destroy-plan graph at all at this point, since we could generate the plan entirely from the state and not need the dependency resolution. We also can't just extract the targeted nodes from the plan graph, because a targeted destroy historically works differently than targeting in other contexts. 

That's all food for future thought though, as this PR is 👍